### PR TITLE
Clone package

### LINKED ISSUE #10: Code review for package_copy
Code review for the three functions in R/package_copy.R would be appreciated!  Open to any general changes / comments.  
https://github.com/NCEAS/datamgmt/blob/master/R/package_copy.R

### PR DESCRIPTION
**Pull request by dmullen17**
Resolve #10  

New version of `clone_package` . 

@amoeba thanks for the review!  I went with most of your comments.  
And sorry it took me so long to reply, editing this function was low on my task list and I just got around to it a few days ago.

### CODE REVIEW

File: R/clone_package.R [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+#' Member node we're cloning from:
+#' cn_from <- dataone::CNode("PROD")
+#' from <- dataone::getMNode(cn_from, "urn:node:ARCTIC")
+#'
+#' Member node we're cloning to:
+#' cn_to <- dataone::CNode('STAGING')
+#' to <- dataone::getMNode(cn_to,'urn:node:mnTestARCTIC')
+#'
+#' Clone the package using a resource map pid:
+#' clone_package(resource_map_doi:10.18739/A2RZ6X", from, to)
```
Comment (amoeba) [inline R/clone_package.R]: Typo, missing a left `"`

Comment (dmullen17) [reply R/clone_package.R]: thanks bryce!

File: R/clone_package.R [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+#' @return (list) Vector of data object pids.
+clone_data_objects <- function(resource_map_pid,
+                               data_pids,
+                               from,
+                               to,
+                               n_max = 3L) {
+    stopifnot(is.character(resource_map_pid))
+    stopifnot(is.character(data_pids))
+    stopifnot(length(data_pids) > 0)
+    stopifnot(methods::is(from, "MNode"))
```
Comment (amoeba) [inline R/clone_package.R]: not sure if you have to explicitly reference `methods::` here. Fine to do so though

**Comment by amoeba** [comment]
Sorry this took me so long. Let me know if you have any questions!

File: R/clone_package.R [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    # Select the characters after the last "/" (the pid)
+    pids <- sapply(urls, function(url) {
+        split_string <- strsplit(url, split = "\\/")
+        utils::tail(split_string[[1]], n = 1)
+    })
+
+    return(pids)
+}
+
+#' Clone data pids between Dataone Member Nodes.
```
[the remark this answers is not in the document]
Comment (dmullen17) [reply R/clone_package.R]: good idea!

File: R/clone_package.R [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+    # Prepare the response object
+    response <- list()
+
+    # Solr query data formatId, fileName, and identifier
+    meta_eml <- get_package_metadata(from, package$resource_map, formatType = "METADATA")
+
+    # Write metadata
+    meta_path <- file.path(tempdir(), meta_eml$fileName)
+    writeBin(dataone::getObject(from, package$metadata), meta_path)
+    new_eml_pid <- arcticdatautils::publish_object(to,
```
[the remark this answers is not in the document]
Comment (dmullen17) [reply R/clone_package.R]: @isteves @emilyodean @maier-m

File: R/clone_package.R (hunk 2 above)
Comment (dmullen17) [reply R/clone_package.R]: Travis might've gotten annoyed at that - not certain though.

File: R/clone_package.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (amoeba) [reply R/clone_package.R]: Yeah, you're right. Also note that the `dataone` package has a similar routine, https://github.com/DataONEorg/rdataone/blob/master/R/D1Object.R#L69. arcticdatautils was written when a few of the classes in the `dataone` package didn't quite work for our use case, but for `datamgmt`, I think many of the classes (DataObject, DataPackage, etc) might work fine

[the remark this answers is not in the document]
Comment (dmullen17) [reply R/clone_package.R]: I think i'm going to leave the calls to `create_resource_map` and `update_resource_map` in, and just update the relevant sysmeta fields immediately after.  

Would these be ok as member node agnostic settings?  
``` 
# ADC: 
slot(sysmeta, "replicationAllowed") <- FALSE
slot(sysmeta, "numberReplicas") <- 0
slot(sysmeta, "blockedNodes") <- list('urn:node:KNB', 'urn:node:mnUCSB1')

# General MN: 
slot(sysmeta, "replicationAllowed") <- NA
slot(sysmeta, "numberReplicas") <- 1
slot(sysmeta, "blockedNodes") <- NA
```

[the remark this answers is not in the document]
Comment (dmullen17) [reply R/clone_package.R]: Ok i'll change it to `generate_resource_map` and `createObject`

**Pull request merged**